### PR TITLE
Empty rolie

### PR DESCRIPTION
**Pull request by JanHoefelmeyer**
Solves https://github.com/csaf-poc/csaf_distribution/issues/336

Csaf checker explicitely filters out if the schema error was no entries and will give a warning for this error only.

Csaf aggregator creates "empty" ROLIE feeds if no entries are present.

Csaf Provider already has an option to create ROLIE feeds during setup. This is currently not mandatory, however, but would be possible to make mandatory if needed.

Csaf Downloader will find feeds with emptry entry, but will not download anything from them.

### CODE REVIEW

File: cmd/csaf_checker/processor.go [hunk 1]
```diff
@@ -677,9 +677,20 @@ func (p *processor) processROLIEFeed(feed string) error {
 		return err
 	}
 	if len(errors) > 0 {
-		p.badProviderMetadata.error("%s: Validating against JSON schema failed:", feed)
-		for _, msg := range errors {
-			p.badProviderMetadata.error(strings.ReplaceAll(msg, `%`, `%%`))
+		lengthError := false
+		for _, lerr := range errors {
+			if lerr == "/feed/entry: minimum 1 items required, but found 0 items" {
```
Comment (s-l-teichmann) [inline cmd/csaf_checker/processor.go:682]: Checking against the text of the specific error message is very end-of-the-pipe and not very robust.

File: cmd/csaf_checker/processor.go [hunk 2]
```diff
@@ -677,9 +677,20 @@ func (p *processor) processROLIEFeed(feed string) error {
...
+				p.badProviderMetadata.warn(lerr)
+				lengthError = true
+			}
+		}
+		if !(lengthError && len(errors) == 2) {
```
Comment (s-l-teichmann) [inline cmd/csaf_checker/processor.go:687]: Checking for a length of 2 feels oddly specific here.

File: cmd/csaf_aggregator/indices.go [hunk 3]
```diff
@@ -157,6 +157,48 @@ func (w *worker) writeIndex(label string, summaries []summary) error {
...
+func (w *worker) writeROLIENoSummaries(label string) error {
+
+	labelFolder := strings.ToLower(label)
+
+	fname := "csaf-feed-tlp-" + labelFolder + ".json"
+
+	feedURL := w.processor.cfg.Domain + "/.well-known/csaf-aggregator/" +
+		w.provider.Name + "/" + labelFolder + "/" + fname
+
+	entries := make([]*csaf.Entry, 0)
```
Comment (s-l-teichmann) [inline cmd/csaf_aggregator/indices.go:169]: nit: Remove this in favor to next comment.

File: cmd/csaf_aggregator/indices.go [hunk 4]
```diff
@@ -157,6 +157,48 @@ func (w *worker) writeIndex(label string, summaries []summary) error {
...
+		Feed: csaf.FeedData{
+			ID:    "csaf-feed-tlp-" + strings.ToLower(label),
+			Title: "CSAF feed (TLP:" + strings.ToUpper(label) + ")",
+			Link:  links,
+			Category: []csaf.ROLIECategory{{
+				Scheme: "urn:ietf:params:rolie:category:information-type",
+				Term:   "csaf",
+			}},
+			Updated: csaf.TimeStamp(time.Now().UTC()),
+			Entry:   entries,
```
Comment (s-l-teichmann) [inline cmd/csaf_aggregator/indices.go:194]: Write `Entry: []*csaf.Entry{},` instead.

**Review by s-l-teichmann — CHANGES_REQUESTED** [review]

File: cmd/csaf_checker/processor.go (hunk 2 above)
Comment (JanHoefelmeyer) [reply cmd/csaf_checker/processor.go:687]: If there is an error due to the length of entry, then another general error will be thrown to indicate failure of validation against the schema. So if there is a length error and *only* the general error, then there is no need to call validation an actual failure. Three or more errors would indicate additional failures.

File: cmd/csaf_checker/processor.go (hunk 1 above)
Comment (JanHoefelmeyer) [reply cmd/csaf_checker/processor.go:682]: csaf.ValidateROLIE (origin of errors) returns only an array of Strings, so at this point it's not considered an error.

File: cmd/csaf_aggregator/indices.go (hunk 4 above)
Comment (JanHoefelmeyer) [reply cmd/csaf_aggregator/indices.go]: Adressed by https://github.com/csaf-poc/csaf_distribution/pull/357/commits/688427798dbc485f1c0a71254036945eb5430365

File: cmd/csaf_checker/processor.go (hunk 1 above)
Comment (JanHoefelmeyer) [reply cmd/csaf_checker/processor.go:682]: Should I then build another test to check whether the entry array is empty? (Since the Checker is supposed to give a warning in that case: https://github.com/csaf-poc/csaf_distribution/issues/336#issuecomment-1421669395)

Comment (tschmidtb51) [reply cmd/csaf_checker/processor.go]: Please use the latest version of the schema from https://github.com/oasis-tcs/csaf/pull/481/files

File: csaf/schema/ROLIE_feed_json_schema.json [hunk 5]
```diff
@@ -36,7 +36,7 @@
           }
         }
       ],
-      "minItems": 1,
+      "minItems": 0,
```
Comment (tschmidtb51) [inline csaf/schema/ROLIE_feed_json_schema.json:39]: Please use the latest version of the schema from https://github.com/oasis-tcs/csaf/pull/481/files

File: cmd/csaf_aggregator/indices.go [hunk 6]
```diff
@@ -311,6 +351,7 @@ func (w *worker) writeService() error {
 func (w *worker) writeIndices() error {
 
 	if len(w.summaries) == 0 || w.dir == "" {
+		w.writeROLIENoSummaries("empty")
```
Comment (tschmidtb51) [inline cmd/csaf_aggregator/indices.go:354]: Hmm. `empty` sounds interesting... Maybe, we should use `undefined`? 
Thoughts?

**Review by tschmidtb51 — CHANGES_REQUESTED** [review]
Please see my comments.

BTW: Shouldn't the change in the `csaf_provider` to generate the appropriated ROLIE feeds be part of this PR or do you want to have that separate?

File: csaf/schema/ROLIE_feed_json_schema.json (hunk 5 above)
Comment (JanHoefelmeyer) [reply csaf/schema/ROLIE_feed_json_schema.json]: Adressed by  https://github.com/csaf-poc/csaf_distribution/pull/357/commits/71de55f7c1ed5f3a579a406d075693023f08bd7c

File: cmd/csaf_checker/processor.go (hunk 1 above)
Comment (JanHoefelmeyer) [reply cmd/csaf_checker/processor.go]: Now uses latest schema version via https://github.com/csaf-poc/csaf_distribution/pull/357/commits/71de55f7c1ed5f3a579a406d075693023f08bd7c .

New function CountEntries now counts entries within a feed and issues a warning should there be 0.

File: cmd/csaf_aggregator/indices.go (hunk 6 above)
Comment (JanHoefelmeyer) [reply cmd/csaf_aggregator/indices.go]: Using undefined now, am open to change if necessary.  (https://github.com/csaf-poc/csaf_distribution/pull/357/commits/2a833bec790477e86c36cd90f43bf260c43c0c3e)

Also: Provider was already able to generate the ROLIE feeds, but the default was set not to. Now the default is to set to generate the feed. (https://github.com/csaf-poc/csaf_distribution/pull/357/commits/418d93bde0ca1a364656a3cee7fcb6f75690c48e)

File: cmd/csaf_checker/processor.go (hunk 2 above)
Comment (JanHoefelmeyer) [reply cmd/csaf_checker/processor.go]: Part has been rewritten

**Review by s-l-teichmann — DISMISSED** [review]
LGTM

**Review by tschmidtb51 — CHANGES_REQUESTED** [review]
1. The creation of empty ROLIE feeds when setting up a provider (using `csaf_uploader -a create`) is not implemented (but it would make the test much easier).
2. Using the aggregator with an empty ROLIE feed (e.g. create a provider, upload one document, delete all entries from the ROLIE feed) results in an error.

Happy to discuss the issues - maybe I misunderstood something

**Comment by mfd2007** [comment]
I can confirm that there is some problem with the aggregator. If the feed is empty, something doesn't work as expected:

[csaf-feed-tlp-white.json.txt](https://github.com/csaf-poc/csaf_distribution/files/11542214/csaf-feed-tlp-white.json.txt)
```
2023/05/23 09:35:19 provider-metadata: https://acme.test/.well-known/csaf/provider-metadata.json
2023/05/23 09:35:19 AdvisoryFileProcessor.Process: Found 1 ROLIE feed(s).
....
2023/05/23 09:35:19 Feed URL: https://acme.test/.well-known/csaf/white/csaf-feed-tlp-white.json
....
2023/05/23 09:35:20 white: 0
....
2023/05/23 09:35:20 error: 'ACME' failed: open /data1/archive/ACME-2023-05-23-093519/white/changes.csv: no such file or directory
```

In that case the provider won't be listed in the aggregator.json. On the other hand, is this a real issue? Does an aggregator want to list an "empty" CSAF provider? Why should the aggregator do this?

According to the provider, I can't find the right option to create the ROLIE feed. For the different TLP-Level there won't be create any feeds.

**Comment by s-l-teichmann** [comment]
> I can confirm that there is some problem with the aggregator. If the feed is empty, something doesn't work as expected:
> 
> [csaf-feed-tlp-white.json.txt](https://github.com/csaf-poc/csaf_distribution/files/11542214/csaf-feed-tlp-white.json.txt)
> 
> ```
> 2023/05/23 09:35:19 provider-metadata: https://acme.test/.well-known/csaf/provider-metadata.json
> 2023/05/23 09:35:19 AdvisoryFileProcessor.Process: Found 1 ROLIE feed(s).
> ....
> 2023/05/23 09:35:19 Feed URL: https://acme.test/.well-known/csaf/white/csaf-feed-tlp-white.json
> ....
> 2023/05/23 09:35:20 white: 0
> ....
> 2023/05/23 09:35:20 error: 'ACME' failed: open /data1/archive/ACME-2023-05-23-093519/white/changes.csv: no such file or directory
> ```
> 
> In that case the provider won't be listed in the aggregator.json. On the other hand, is this a real issue? Does an aggregator want to list an "empty" CSAF provider? Why should the aggregator do this?
> 
> According to the provider, I can't find the right option to create the ROLIE feed. For the different TLP-Level there won't be create any feeds.

The error in the aggregator is related to the fact that there are
no entries. So writing 'changes.csv' and 'index.txt' will fail as the respective folders do not exist.

commit 7729a83d02c463e9191bc3f8dd856c61d99535d6 stops before writing these files so the error should not occur any more.

**Review by s-l-teichmann — APPROVED** [review]
Code looks good to me but I really would like to have @tschmidtb51 and/or @mfd2007 have a look at the feature expectation side.

**Comment by tschmidtb51** [comment]
Sure. Sorry for the delay. I'll try to squeeze the review in until end of this week.

File: csaf/rolie.go [hunk 7]
```diff
@@ -238,3 +238,8 @@ func (rf *ROLIEFeed) SortEntriesByUpdated() {
 		return time.Time(entries[j].Updated).Before(time.Time(entries[i].Updated))
 	})
 }
+
+// CountEntries returns the number of entries within the feed
+func (rf *ROLIEFeed) CountEntries() int {
+	return len(rf.Feed.Entry)
```
Comment (tschmidtb51) [inline csaf/rolie.go:244]: I guess this function fails, if a feed does not contain entries.

File: cmd/csaf_provider/create.go [hunk 8]
```diff
@@ -153,10 +154,55 @@ func createFeedFolders(c *config, wellknown string) error {
...
+	rolie := &csaf.ROLIEFeed{
+		Feed: csaf.FeedData{
+			ID:    "csaf-feed-tlp-" + ts,
+			Title: "CSAF feed (TLP:" + string(tlpLabel) + ")",
+			Link:  links,
+			Category: []csaf.ROLIECategory{{
+				Scheme: "urn:ietf:params:rolie:category:information-type",
+				Term:   "csaf",
+			}},
+			Updated: csaf.TimeStamp(time.Now().UTC()),
```
Comment (tschmidtb51) [inline cmd/csaf_provider/create.go:198]: The `entry` array should be added here as empty array: `Entry: []*csaf.Entry{},`

**Review by tschmidtb51 — CHANGES_REQUESTED** [review]
The `csaf_provider` currently does create an empty ROLIE feed but the `entry` array is missing. If I run the `csaf_checker` on that, I get a segmentation violation. Also the `csaf_aggregator` fails the with missing ROLIE feed (`Example_Company_02-2023-06-16-075222/white/csaf-feed-tlp-white.json: no such file or directory`). The `csaf_downloader` does not have an issue with that situation.

If I add the `entry` array to the feed, then:
- `csaf_checker` still results in a segmentation violation (`main.go:174`->`main.go:151`->`processor.go:239`->`processor.go:319`->`processor.go:977`->`roilecheck.go:247`)
- `csaf_downloader`still works fine
- `csaf_aggregator` still fails (so this might be unrelated)

**Review by tschmidtb51 — CHANGES_REQUESTED** [review]
I can confirm that the checker is now running fine (in all tested conditions).

The `csaf_provider` still needs to create the `entry` array. (see https://github.com/csaf-poc/csaf_distribution/pull/357#discussion_r1231888352)

I haven't located the bug in the aggregator why it is not creating a ROLIE feed with an empty feed array, but I'm happy to document that as it is consistent (doesn't show up in the `aggregator.json` either) - and leave that for now.

File: cmd/csaf_provider/create.go (hunk 8 above)
Comment (s-l-teichmann) [reply cmd/csaf_provider/create.go:198]: Addressed in c4950065475d4314efb5f171cfad16435940bed2

**Review by s-l-teichmann — DISMISSED** [review]
LGTM

**Review by mfd2007 — DISMISSED** [review]
LGTM

File: cmd/csaf_checker/processor.go [hunk 9]
```diff
@@ -445,12 +450,15 @@ func (p *processor) rolieFeedEntries(feed string) ([]csaf.AdvisoryFile, error) {
...
 		return rfeed, rolieDoc, err
-
 	}()
 	if err != nil {
 		p.badProviderMetadata.error("Loading ROLIE feed failed: %v.", err)
 		return nil, errContinue
 	}
+
+	if rfeed.CountEntries() == 0 {
+		p.badProviderMetadata.warn("No entries in %s", feed)
```
Comment (tschmidtb51) [inline cmd/csaf_checker/processor.go:460]: Doesn't it make more sense to report that under the ROLIE feeds? (Requirement 15)? Just a question - if you agree, please change it; if not please explain.

File: cmd/csaf_checker/processor.go [hunk 10]
```diff
@@ -251,6 +251,11 @@ func (p *processor) run(domains []string) (*Report, error) {
 			continue
 		}
 
+		if domain.Role == nil {
+			log.Printf("No role found in meta data. Ignoring domain %q\n", d)
```
[the remark this answers is not in the document]
Comment (tschmidtb51) [reply cmd/csaf_checker/processor.go:255]: BTW: The same error shows up, if the `/distributions[]/rolie/feeds[]`is empty (but the file has a valid `role`).

To reproduce: Setup a new `csaf_provider`. Delete all entries from `/distributions[]/rolie/feeds[]` in the PMD. Run the `csaf_checker`.

**Review by tschmidtb51 — CHANGES_REQUESTED** [review]
Please see my comments. Additional remarks:

1. I couldn't test the `csaf_aggreagtor`change because #377. It fails as soon as you have a ROLIE feed with an empty `entry` array. See example error message: `'Example_Company_02' failed: open /var/www/aggregator-data/Example_Company_02-2023-06-23-222130/white/csaf-feed-tlp-white.json: no such file or directory`.
2. There is still some confusion regarding the error message and its meaning in the `csaf_checker.`
3. The change in https://github.com/csaf-poc/csaf_distribution/commit/c4950065475d4314efb5f171cfad16435940bed2 didn't fix the issue in the `csaf_provider`. The `entry` array is still empty when delivered.
4.

File: cmd/csaf_checker/processor.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (JanHoefelmeyer) [reply cmd/csaf_checker/processor.go:257]: This all seems to have the same root cause: There is no proper Error handling anymore if the checkProviderMetadata() function fails. 

All these tests actually cause a validation error against the JSON schema, which causes domain to only have a name, and as such, always cause domain.Role == nil.

https://github.com/csaf-poc/csaf_distribution/pull/357/commits/7a638329f6841fe36b23f8c8d6b94fc5065071ba addresses this.

File: cmd/csaf_checker/processor.go (hunk 9 above)
Comment (JanHoefelmeyer) [reply cmd/csaf_checker/processor.go]: I agree, adressed by https://github.com/csaf-poc/csaf_distribution/pull/357/commits/f7ed10af7747d843aba87fa84f0252c4a6d68971

File: cmd/csaf_checker/processor.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (tschmidtb51) [reply cmd/csaf_checker/processor.go:258]: Should work for now. Additional improvement suggested through #385 and #386 (out-of-scope for 2.2.0)

**Comment by tschmidtb51** [comment]
@JanHoefelmeyer https://github.com/csaf-poc/csaf_distribution/pull/357#discussion_r1240434944 still exists - the other two have been resolved with the last 3 commits.

There are also merge conflicts...

File: cmd/csaf_provider/create.go (hunk 8 above)
Comment (s-l-teichmann) [reply cmd/csaf_provider/create.go:198]: commit 26a4d2e7b94f0564f7d3a27ee419fb004e888392 removes the `omitempty` rule which eliminates empty elements from the JSON result. This should finally help.

**Review by tschmidtb51 — CHANGES_REQUESTED** [review]
The merge conflict resolution seems to be gone wrong - now any domain requested fails. I haven't found the line number yet where the issue is. But basically, it reports 
```
2023/06/30 20:43:10 Failed to find valid provider-metadata.json for domain domain.tld. Continuing with next domain.
{
  "version": "2.1.1-123-g30c2218",
  "date": "2023-06-30T18:43:02.567470825Z"
}
````
for any valid domain

File: cmd/csaf_provider/create.go (hunk 8 above)
Comment (tschmidtb51) [reply cmd/csaf_provider/create.go:198]: This is resolved.

**Comment by s-l-teichmann** [comment]
> The merge conflict resolution seems to be gone wrong - now any domain requested fails. I haven't found the line number yet where the issue is. But basically, it reports
> 
> ```
> 2023/06/30 20:43:10 Failed to find valid provider-metadata.json for domain domain.tld. Continuing with next domain.
> {
>   "version": "2.1.1-123-g30c2218",
>   "date": "2023-06-30T18:43:02.567470825Z"
> }
> ```
> 
> for any valid domain

@tschmidtb51 That was indeed a problem introduced by a faulty merge.
(Forgot to remove an `else` which should not be there)

 I cleaned up the error handling in commit 89e69a7bafd93efa9c6fd24d7b49241874ddb9cf .

**Comment by s-l-teichmann** [comment]
Reopen

**Review by tschmidtb51 — APPROVED** [review]
LGTM and works

**Comment by tschmidtb51** [comment]
@s-l-teichmann Thank you for the fast fix!

**Review by s-l-teichmann — APPROVED** [review]
LGTM
